### DISCUSSION TITLE
ReDos vulnerability back?

### DISCUSSION THREAD

**Discussion by CrossEye** [ramda/ramda#3286] [General]
@customcommander, @adispring, @kedashoe, @semmel:

Anyone have an idea why #3177 is back?  We're getting reports about it.

I would just dismiss it out of hand, as this:

```js
const {trim} = require ('ramda');

function build_blank (n) {
  var ret = "1"
  for (var i = 0; i < n; i++) {
    ret += " "
  }

  return ret + "1";
}

var time = Date.now();
trim(build_blank (50000))
var time_cost = Date.now() - time;
console.log("time_cost: " + time_cost)
```

runs in a few milliseconds in Node, under `0.27.2`, under `0.28.0`, and under `latest` npm installs.

But it seems to be still a problem [in the REPL][rr], taking around three seconds for me.  It's hard to tell with the minimized code if the regex character removed is still missing, but the other tweak in the relevant commit is still there, so I assume it is.  

I'm confused.  Any ideas?

  [rr]: https://ramdajs.com/repl/#?console%20.clear%20%28%29%0A%0Afunction%20build_blank%20%28n%29%20%7B%0A%20%20var%20ret%20%3D%20%221%22%0A%20%20for%20%28var%20i%20%3D%200%3B%20i%20%3C%20n%3B%20i%2B%2B%29%20%7B%0A%20%20%20%20ret%20%2B%3D%20%22%20%22%0A%20%20%7D%0A%0A%20%20return%20ret%20%2B%20%221%22%3B%0A%7D%0A%0Avar%20time%20%3D%20Date.now%28%29%3B%0Atrim%28build_blank%20%2850000%29%29%0Avar%20time_cost%20%3D%20Date.now%28%29%20-%20time%3B%0Aconsole.log%28%22time_cost%3A%20%22%20%2B%20time_cost%29

**Comment by adispring** [comment]
I did some experiments with `https://cdn.jsdelivr.net/npm/ramda@latest/dist/ramda.min.js`(which is used in ramda repl), in node and web browser (chrome/safari). I found native `String.prototype.trim` in browser is slow, and this cause `R.trim` slow, because `R.trim` will dispatch to `String.prototype.trim` if possible.

Here is some test code and result:

1. In node.js on my macbook:
```js
  console.clear();

  function build_blank(n) {
    var ret = '1';
    for (var i = 0; i < n; i++) {
      ret += ' ';
    }

    return ret + '1';
  }

  // console.log(t, R);
  var time = Date.now();
  t.trim(build_blank(50000));
  var time_cost = Date.now() - time;
  console.log('time_cost: ' + time_cost);

  var time2 = Date.now();
  build_blank(50000).trim();
  var time_cost2 = Date.now() - time2;
  console.log('time_cost2: ' + time_cost2);
```
result:
time_cost: 36
time_cost2: 3


2. In chrome:
```js
  console.clear();

  function build_blank(n) {
    var ret = '1';
    for (var i = 0; i < n; i++) {
      ret += ' ';
    }

    return ret + '1';
  }

  // console.log(t, R);
  var time = Date.now();
  R.trim(build_blank(50000));
  var time_cost = Date.now() - time;
  console.log('time_cost: ' + time_cost);

  var time2 = Date.now();
  build_blank(50000).trim();
  var time_cost2 = Date.now() - time2;
  console.log('time_cost2: ' + time_cost2);
```
result:
time_cost: 3115
time_cost2: 3157

3. In safari with the same test code as in chrome
result:
time_cost: 2236
time_cost2: 2207

**Reply by semmel** [reply]
Don't include the time to generate the test string in your measurements. That loop takes way more time than the trimming operation.

**Reply by CrossEye** [reply]
Although that makes sense, and I've done some of that in my own testing, this is the test case as given by the vulnerability report.

I wonder if it simply makes sense to stop defaulting to `String.prototype.trim`.  I'm not really sure why we did that in the first place unless we were worried about not getting the space characters right.

**Reply by semmel** [reply]
> I wonder if it simply makes sense to stop defaulting to String.prototype.trim

I will include Regex trimming in my test demo in my other reply in this issue asap. 
However, as I am repeating here: String.prototype.trim is not slower than R.trim. It makes no sense to measure the time to execute the for loop. Cannot draw conclusions based on faulty assumptions.

**Reply by semmel** [reply]
@adispring 
> I found native String.prototype.trim in browser is slow, and this cause R.trim slow, because R.trim will dispatch to String.prototype.trim if possible.

If that would be the case, "Prisma" should mark all JS browser engines as "vulnerable".

@CrossEye 
> I wonder if it simply makes sense to stop defaulting to `String.prototype.trim`

On the contrary; we should remove the regex fallback entirely and only dispatch to `String.prototype.trim`. The regex method is horribly slow as my tests (https://jsitor.com/448sTwDhQ) show (at least 500 x slower). It only takes effect on Internet Explorer 8 or Node.js < v0.1. Which makes me wonder how "Prisma" came up with their CVE.

**Comment by semmel** [comment]
I cannot confirm it, to trim a 50.000.000 char string on our 2013 Macbook , Chrome takes
943msec using Ramda
602msec using native `String.trim`.

Makes sure you don't include the string generation in your time measurements.
https://jsitor.com/448sTwDhQ

Edit 2.6.2022
Added Ramda's trimming fallback implementation based on regex to the measurements. Turns out it is at least factor 500 x slower than `String.prototype.trim`.
